controller/src/main/java/io/pravega/controller/store/client/StoreClientFactory.java

```diff
@@ -108,19 +117,31 @@ public ZooKeeper newZooKeeper(String connectString, int sessionTimeout, Watcher
             if (client == null) {
                 Exceptions.checkNotNullOrEmpty(connectString, "connectString");
                 Preconditions.checkArgument(sessionTimeout > 0, "sessionTimeout should be a positive integer");
-                Preconditions.checkNotNull(watcher, "watcher");
                 this.connectString = connectString;
                 this.sessionTimeout = sessionTimeout;
                 this.canBeReadOnly = canBeReadOnly;
                 this.client = new ZooKeeper(connectString, sessionTimeout, watcher, canBeReadOnly);
-                return this.client;
             } else {
-                Preconditions.checkArgument(this.connectString.equals(connectString), "connectString differs");
-                Preconditions.checkArgument(this.sessionTimeout == sessionTimeout, "sessionTimeout differs");
-                Preconditions.checkArgument(this.canBeReadOnly == canBeReadOnly, "canBeReadOnly differs");
-                Preconditions.checkNotNull(watcher, "watcher");
-                this.client.register(watcher);
-                return this.client;
+                try {
+                    Preconditions.checkArgument(this.connectString.equals(connectString), "connectString differs");
+                    Preconditions.checkArgument(this.sessionTimeout == sessionTimeout, "sessionTimeout differs");
+                    Preconditions.checkArgument(this.canBeReadOnly == canBeReadOnly, "canBeReadOnly differs");
+                    this.client.register(watcher);
+                } catch (IllegalArgumentException e) {
+                    log.warn("Input argument for new ZooKeeper client ({}, {}, {}) changed with respect to existing client ({}, {}, {}).",
+                        connectString, sessionTimeout, canBeReadOnly, this.connectString, this.sessionTimeout, this.canBeReadOnly, e);
+                    closeClient(client);
+                }
+            }
+            return this.client;
+        }
+
+        private void closeClient(ZooKeeper client) {
+            try {
+                client.close();
+            } catch (Exception e) {
+                // We prevent throwing uncontrolled exceptions here, which may lead Curator to retry indefinitely.
+                log.error("Problems attempting to close ZooKeeper client.", e);
```

Comment:
Say instead:

"Exception while attempting to close ZooKeeper client."

Reply:
Done.